### PR TITLE
Add tests for MNIST data loader and update data loading functionality

### PR DESCRIPTION
**Pull request by ketayon**
This PR addresses issue #101 by adding train_test_split functionality within the load_mnist_dataset function. Now, users no longer need to manually split the MNIST dataset into training and testing sets before loading.

Changes:

1. Added train-test splitting based on the split_ratio argument, allowing for flexible dataset sizes.
2. Updated the function to allow loading only the train or test dataloader, or both, enhancing usability. 
3. Added comprehensive tests in test_mnist_data_loader.py to ensure the functionality of the load_mnist_dataset function, including:

- Verifying that the dataset splits correctly according to the specified split_ratio.
- Testing the functionality of loading train and test dataloaders independently or together.
- Checking that the data normalization functionality works as intended.

Let me know if any further changes are required.

### CODE REVIEW

**Comment by SaashaJoshi** [comment]
Hi @ketayon, thank you for the contribution. I will get back to reviewing as soon as I can. However, are you in a rush to get this done before hacktoberfest gets over?

**Comment by ketayon** [comment]
Hello, no, I don’t take part in hacktoberfest.On 22 Oct 2024, at 20:49, Saasha Joshi ***@***.***> wrote:Hi @ketayon, thank you for the contribution. I will get back to reviewing as soon as I can. However, are you in a rush to get this done before hacktoberfest gets over?

File: piqture/data_loader/mnist_data_loader.py [hunk 1]
```diff
@@ -13,123 +13,130 @@
...
-                torchvision.transforms.ToTensor(),
-                torchvision.transforms.Resize(img_size),
-            ]
-        )
-
-    # Partial application of custom collate function that calls
-    # collate_fn with args.
-    new_batch = []
-    custom_collate = partial(collate_fn, labels=labels, new_batch=new_batch)
+        raise TypeError("img_size tuple must contain integers.")
```
Comment (SaashaJoshi) [inline piqture/data_loader/mnist_data_loader.py:61]: Can we revert the TypeError to what it was before? I prefer having more details in the error message.

File: piqture/data_loader/mnist_data_loader.py [hunk 2]
```diff
@@ -13,123 +13,130 @@
...
-def load_mnist_dataset(
-    img_size: Union[int, tuple[int, int]] = 28,
-    batch_size: int = None,
-    labels: list = None,
-    normalize_min: float = None,
-    normalize_max: float = None,
+def load_mnist_dataset(  # pylint: disable=R0913, R0914, W0012
+    img_size: Union[int, Tuple[int, int]] = 28,
+    batch_size_train: int = 64,
+    batch_size_test: int = 1000,
```
Comment (SaashaJoshi) [inline piqture/data_loader/mnist_data_loader.py:28]: Having separate `batch_size` arguments is good!

Comment (SaashaJoshi) [reply piqture/data_loader/mnist_data_loader.py:28]: However, any particular reason behind these arguments having default values of 64 and 1000?

File: piqture/data_loader/mnist_data_loader.py [hunk 3]
```diff
@@ -13,123 +13,130 @@
...
+        labels (list, optional): List of desired labels.
         normalize_min (float, optional): Minimum value for normalization.
         normalize_max (float, optional): Maximum value for normalization.
+        split_ratio (float, optional): Ratio to split train/test datasets. Defaults to 0.8.
+        load (str, optional): Indicates whether to load "train", "test", or "both".
+        Defaults to "both".
 
     Returns:
-        Train and Test DataLoader objects.
+        Train and/or Test DataLoader objects, depending on the `load` argument.
```
Comment (SaashaJoshi) [inline piqture/data_loader/mnist_data_loader.py:51]: Can we say,
```suggestion
        Train and/or Test DataLoader objects, depending on the `dataset` argument.
```

File: piqture/data_loader/mnist_data_loader.py [hunk 4]
```diff
@@ -13,123 +13,130 @@
...
-    # collate_fn with args.
-    new_batch = []
-    custom_collate = partial(collate_fn, labels=labels, new_batch=new_batch)
+        raise TypeError("img_size tuple must contain integers.")
+
+    if not isinstance(batch_size_train, int) or not isinstance(batch_size_test, int):
+        raise TypeError("batch_size_train and batch_size_test must be integers.")
+
+    if labels is not None and not isinstance(labels, list):
+        raise TypeError("labels must be a list.")
```
Comment (SaashaJoshi) [inline piqture/data_loader/mnist_data_loader.py:67]:
```suggestion
        raise TypeError("The input labels must be of the type list.")
```

File: piqture/data_loader/mnist_data_loader.py [hunk 5]
```diff
@@ -13,123 +13,130 @@
...
-    normalize_max: float = None,
+def load_mnist_dataset(  # pylint: disable=R0913, R0914, W0012
+    img_size: Union[int, Tuple[int, int]] = 28,
+    batch_size_train: int = 64,
+    batch_size_test: int = 1000,
+    labels: Optional[List[int]] = None,
+    normalize_min: Optional[float] = None,
+    normalize_max: Optional[float] = None,
+    split_ratio: float = 0.8,
+    load: str = "both",  # Options: "train", "test", or "both"
```
Comment (SaashaJoshi) [inline piqture/data_loader/mnist_data_loader.py:33]:
```suggestion
    dataset: str = "both",  # Options: "train", "test", or "both"
```

File: piqture/data_loader/mnist_data_loader.py [hunk 6]
```diff
@@ -13,123 +13,130 @@
...
+        raise TypeError("img_size tuple must contain integers.")
+
+    if not isinstance(batch_size_train, int) or not isinstance(batch_size_test, int):
+        raise TypeError("batch_size_train and batch_size_test must be integers.")
+
+    if labels is not None and not isinstance(labels, list):
+        raise TypeError("labels must be a list.")
+
+    if load not in {"train", "test", "both"}:
+        raise ValueError('load must be one of "train", "test", or "both".')
```
Comment (SaashaJoshi) [inline piqture/data_loader/mnist_data_loader.py:70]:
```suggestion
    if dataset not in {"train", "test", "both"}:
        raise ValueError('The dataset argument must be one of "train", "test", or "both".')
```

File: piqture/data_loader/mnist_data_loader.py [hunk 7]
```diff
@@ -13,123 +13,130 @@
...
+    # Define the transform
+    mnist_transform = torchvision.transforms.Compose(
+        [
+            torchvision.transforms.ToTensor(),
+            torchvision.transforms.Resize(img_size),
+            (
+                MinMaxNormalization(normalize_min, normalize_max)
+                if normalize_min is not None
+                and normalize_max is not None  # pylint: disable=C0301
+                else torchvision.transforms.Lambda(lambda x: x)
```
Comment (SaashaJoshi) [inline piqture/data_loader/mnist_data_loader.py:81]: We do not need an else statement.

File: piqture/data_loader/mnist_data_loader.py [hunk 8]
```diff
@@ -13,123 +13,130 @@
...
-    # Download dataset.
-    mnist_train = datasets.MNIST(
+    # Load the full MNIST dataset
+    mnist_full = datasets.MNIST(
         root="data/mnist_data",
-        train=True,
+        train=True,  # Always load train to split later
         download=True,
         transform=mnist_transform,
     )
```
Comment (SaashaJoshi) [inline piqture/data_loader/mnist_data_loader.py:92]: PyTorch suggests to load train and test datasets separately. I would prefer keeping it the original way. 

```python
# Download dataset.
mnist_train = datasets.MNIST(
    root="data/mnist_data",
    train=True,
    download=True,
    transform=mnist_transform,
)

mnist_test = datasets.MNIST(
    root="data/mnist_data", train=False, download=True, transform=mnist_transform
)
```

**Comment by SaashaJoshi** [comment]
Hi @ketayon,

I have been reviewing this PR and this made me realize that I would prefer loading the train and test datasets separately, as PyTorch suggets. Like here, https://github.com/SaashaJoshi/piQture/blob/b3e7203cadc8b95e2e8ec00b8ccc1ffa5825cf97/piqture/data_loader/mnist_data_loader.py#L93-L103

With this, I would like to add a `ratio` argument to mention how many `train:test` samples a user wishes to load. This argument would provide a value to the `batch_size` argument. 

However, the idea of just loading the train set and splitting it into train and test dataset is not bad, but it unnecessarily reduces the  total amount of data we load. Do you think this is correct? Let me know if you think otherwise.
